File: python/structural0.py
from abc import ABCMeta, abstractmethod
from math import sqrt
from typing import Union, Protocol, SupportsRound

class Rectangle:
    def __init__(
        self,
        width: int,
        height: int,
    ):
        self.width = width
        self.height = height

    def area(self) -> int:
        return self.width * self.height


class EquilateralTriangle:
    def __init__(self, side_len: int):
        self.side_len = side_len

    def area(self) -> float:
        return sqrt(3) / 4 * self.side_len ** 2

class SwimmingPool:
    def area(self) -> float:
        return 50.5


class Shape(Protocol):
    def area(self) -> SupportsRound: ...

def round_area(shape: Shape):
    return int(round(shape.area()))


round_area(Rectangle(100, 50))
round_area(EquilateralTriangle(5))
round_area(SwimmingPool())

File: typescript/structural0.ts

class Rectangle {
  constructor(private x: number, private y: number) { }
  area = () => this.x * this.y
  height = () => Math.max(this.x, this.y)
}

class EquilateralTriangle {
  constructor(private side_len: number) { }
  area = () => Math.sqrt(3) / 4 * this.side_len ** 2
  height = () => (this.side_len * Math.sqrt(3)) * 2
}

export const roundArea = (shape: { area(): number }) => Math.round(shape.area())


class SwimmingPool {
  area = () => 55.5
}

roundArea(new Rectangle(100, 50))
roundArea(new EquilateralTriangle(5))
roundArea(new SwimmingPool())
roundArea({ area: () => 20.8 })
